connect_interceptor.go

```diff
@@ -0,0 +1,56 @@
+package mssql
+
+import "context"
+
+// contextKey is, as the name implied, a type reserved
+// for keys when passing values into the context
+type contextKey string
+const connectInterceptorKey contextKey = "connectInterceptor"
+
+// ConnectInterceptor is used to exchange values between the driver and the user during
+// the connection phase
+type ConnectInterceptor struct {
+	// PreLoginResponse is used to obtain PreLogin Response fields
+	PreLoginResponse chan map[uint8][]byte
+	// ClientLogin is used to pass the client Login
+	ClientLogin chan Login
+	// ServerLoginAck is used to pass the server LoginAck
+	ServerLoginAck chan LoginAckStruct
+	// ServerError is used to pass the server error
+	ServerError chan Error
+}
```

Comment:
I really like this abstraction. 

Not a blocker, but I would consider a simpler name like `ConnectionSteps` or `ProtocolSteps` or `ProtocolStepResponses` (etc).    Going back to the conceptual purity argument that convinced me to put this stuff here: If this is part of an extended public API for the driver, the fact that we're hooking into the `Connect` method is an impl detail we don't care about.  

It's really providing us a way to get the responses from the various protocol steps.